mailpoet/lib/Automation/Engine/Control/SubjectTransformerHandler.php

```diff
@@ -0,0 +1,142 @@
+<?php declare(strict_types = 1);
+
+namespace MailPoet\Automation\Engine\Control;
+
+use MailPoet\Automation\Engine\Data\Automation;
+use MailPoet\Automation\Engine\Data\AutomationRun;
+use MailPoet\Automation\Engine\Data\Step as StepData;
+use MailPoet\Automation\Engine\Data\Subject as SubjectData;
+use MailPoet\Automation\Engine\Integration\Step;
+use MailPoet\Automation\Engine\Integration\SubjectTransformer;
+use MailPoet\Automation\Engine\Registry;
+use MailPoet\Automation\Engine\Storage\AutomationStorage;
+
+class SubjectTransformerHandler {
+
+  /* @var Registry */
+  private $registry;
+
+  /* @var AutomationStorage */
+  private $automationStorage;
+
+  public function __construct(
+    Registry $registry,
+    AutomationStorage $automationStorage
+  ) {
+    $this->registry = $registry;
+    $this->automationStorage = $automationStorage;
+  }
+
+  /** @return string[] */
+  public function subjectKeysForAutomation(Automation $automation): array {
+    $configuredTriggers = array_filter(
+      $automation->getSteps(),
+      function (StepData $step): bool {
+        return $step->getType() === StepData::TYPE_TRIGGER;
+      }
+    );
+
+    $triggers = array_filter(array_map(
+      function(StepData $step): ?Step {
+        return $this->registry->getStep($step->getKey());
+      },
+      $configuredTriggers
+    ));
```

Comment:
Imo save as $triggerData was wrapped in array_values, if I understand you correctly.

Reply:
The `$triggers` array can become associative due to `array_filter`, but in the current code, it's not a problem. Now I'm actually more thinking about what if a trigger disappears due to some plugin being temporarily deactivated, but I hope that's OK because in the UI, we'd already display an "unknown" trigger, in that case.

Reply:
`public function getSubjectKeysForAutomation()` -> Used in the validator: The result could be the Automation is invalid when you try to update it, but imo that should be already the case as the trigger is missing. Not sure, but rather an effect we would want to have.

`public function getSubjectKeysForTrigger(Trigger $trigger)` -> Used to give the information, which subjects are supported to the React applicaton. No effect as `Trigger $trigger` does not exist.

`public function getAllSubjects(array $subjects)` -> Used when collecting all subjects when the trigger was triggered. We return all subjects of this (not deactivated) trigger and what we can derive from this trigger. Thanks to `array_intersect` when validating the automation the returns subjects will be sufficient for all following steps.

TLDR;
> I hope that's OK because

I do not see a problem.